client/src/annoMatrix/views.ts

```diff
@@ -4,37 +4,59 @@
 Views on the annomatrix.  all API here is defined in viewCreators.js and annoMatrix.js.
 */
 import clip from "../util/clip";
-import AnnoMatrix from "./annoMatrix";
-import { _whereCacheCreate } from "./whereCache";
+import AnnoMatrix, { ObsColumnValue } from "./annoMatrix";
+import { _whereCacheCreate, WhereCache } from "./whereCache";
 import { _isContinuousType, _getColumnSchema } from "./schema";
+import {
+  Dataframe,
+  DenseInt32Index,
+  IdentityInt32Index,
+  KeyIndex,
+} from "../util/dataframe";
+import { Query } from "./query";
+import {
+  AnnotationColumn,
+  ColumnSchema,
+  Field,
+  LayoutColumn,
+} from "../common/types/entities";
 
-class AnnoMatrixView extends AnnoMatrix {
-  // eslint-disable-next-line @typescript-eslint/no-explicit-any --- FIXME: disabled temporarily on migrate to TS.
-  constructor(viewOf: any, rowIndex = null) {
-    // @ts-expect-error ts-migrate(2531) FIXME: Object is possibly 'null'.
+type MapFn = (
+  field: Field,
+  colLabel: string,
+  colSchema: ColumnSchema,
+  colData: string[],
+  df: Dataframe
+) => string[];
+
+abstract class AnnoMatrixView extends AnnoMatrix {
+  constructor(
+    viewOf: AnnoMatrix,
+    rowIndex: DenseInt32Index | IdentityInt32Index | KeyIndex | null = null
+  ) {
     const nObs = rowIndex ? rowIndex.size() : viewOf.nObs;
     super(viewOf.schema, nObs, viewOf.nVar, rowIndex || viewOf.rowIndex);
     this.viewOf = viewOf;
     this.isView = true;
   }
 
-  // eslint-disable-next-line @typescript-eslint/no-explicit-any --- FIXME: disabled temporarily on migrate to TS.
-  addObsAnnoCategory(col: any, category: any) {
+  addObsAnnoCategory(col: string, category: string): AnnoMatrix {
     const newAnnoMatrix = this._clone();
+    // @ts-expect-error ts-migrate --- TODO revisit:
+    // Compiler is complaining that this.viewOf is possibly null, which is true when the AnnoMatrix is the base. Consider
```

Comment:
I think the actual issue here is that the type of `newAnnoMatrix` is `AnnoMatrix`, rather than `AnnoMatrixView`.  This is inherited from the base-class-implemented clone() method.   Maybe clone() should be sub-class implemented, _or_ just cast it...

Reply:
Maybe I am missing something here! I feel the core problem is that `viewOf` is optional and can possibly have a value of `undefined` regardless of the instance type. I'm not sure casting `newAnnoMatrix` to `AnnoMatrixView` will resolve the compiler errors. For example, I am still getting an ` Object is possibly 'undefined'.` on `this.viewOf` in `this.viewOf.addObsAnnoCategory` with the following:

```
const newAnnoMatrix = this._clone() as AnnoMatrixView;
newAnnoMatrix.viewOf = this.viewOf.addObsAnnoCategory(col, category);
```

Reply:
yep,I think you are right.

A couple of possibilities:
* we could make `viewOf` required, and simply has the base (loader) view point to itself (make sure `isView === false`)
* you could add type narrowing in these functions.

they are equivalent IMHO, although one requires far less code.

Reply:
Updated [`viewOf` to be required](https://github.com/chanzuckerberg/cellxgene/blob/mim/annoMatrix/client/src/annoMatrix/annoMatrix.ts#L71) and [set initial value to self](https://github.com/chanzuckerberg/cellxgene/blob/mim/annoMatrix/client/src/annoMatrix/annoMatrix.ts#L148).

Reply:
Comment on L36 of annoMatrix.ts still needs to be updated to match this change.

Reply:
Also: I'm not sure you updated all of the locations that use `viewOf` to understand this new protocol.  An example: middleware.ts:L66 (will now get into an infinite loop).

Reply:
Fixed defect in [middleware:L72](https://github.com/chanzuckerberg/cellxgene/blob/mim/annoMatrix/client/src/annoMatrix/middleware.ts#L72) and updated [`viewOf` tsdoc](https://github.com/chanzuckerberg/cellxgene/blob/mim/annoMatrix/client/src/annoMatrix/annoMatrix.ts#L139).

Reply:
I like the most recent refinement...